llm/cpp/llm.cpp

```diff
@@ -3,30 +3,332 @@
+                for (Beam& beam : groups[group_idx].beams) {
+                    beam.tokens.push_back(pad_token);
+                }
+                continue;
+            }
+            std::vector<Beam> candidates;
+            candidates.reserve(2 * GROUP_SIZE);
+            for (size_t beam_idx = 0; beam_idx < GROUP_SIZE; ++beam_idx) {
+                if (logits.get_shape()[0] <= global_beam_ids[group_idx][beam_idx]) {
+                    throw std::runtime_error("logits batch size doesn't match the number of beams");
+                }
+                size_t vocab_size = logits.get_shape().back();
+                std::vector<float> temp;
+                size_t batch_offset = global_beam_ids[group_idx][beam_idx] * logits.get_shape()[1] * logits.get_shape()[2];
+                const float* beam_logits = logits.data<const float>() + batch_offset + (logits.get_shape()[1] - 1) * vocab_size;
+                float max_logit = *std::max_element(beam_logits, beam_logits + vocab_size);
+                float log_sum = std::log(std::accumulate(beam_logits, beam_logits + vocab_size, 0.0f, [max_logit](float accumulated, float to_add) {
+                    return accumulated + std::exp(to_add - max_logit);
+                }));
+                std::vector<Token> tokens;
+                tokens.reserve(vocab_size);
+                for (size_t idx = 0; idx < vocab_size; ++idx) {
+                    tokens.push_back({beam_logits[idx] - max_logit - log_sum, int64_t(idx)});
+                }
+                for (size_t prev_group_idx = 0; prev_group_idx < group_idx; ++prev_group_idx) {  // TODO: range based for
+                    for (size_t prev_beam_idx = 0; prev_beam_idx < GROUP_SIZE; ++prev_beam_idx) {
+                        tokens[size_t(groups[prev_group_idx].beams[prev_beam_idx].tokens.back())].log -= diversity_penalty;
+                    }
+                }
+                std::vector<int64_t>& other_tokens = groups[group_idx].beams[beam_idx].tokens;
+                std::vector<int64_t> full_text;
+                for (size_t idx = 0; idx < input_ids.get_size(); ++idx) {
+                    full_text.push_back(input_ids.data<int64_t>()[idx]);
+                }
+                full_text.insert(full_text.end(), other_tokens.begin(), other_tokens.end());
+                if (full_text.size() > 1 && full_text.size() >= no_repeat_ngram_size) {
+                    for (int64_t banned_token : kmp_search(full_text, {full_text.end() - ptrdiff_t(no_repeat_ngram_size) + 1, full_text.end()})) {
+                        tokens[size_t(banned_token)].log = -std::numeric_limits<float>::infinity();
+                    }
+                }
+                std::sort(tokens.begin(), tokens.end());
+                size_t new_token_idx = 0;
+                for (int added_count = 0; added_count < int(2 * GROUP_SIZE); ++added_count) {
+                    candidates.push_back(groups[group_idx].beams[beam_idx]);
```

Comment:
maybe we can create `Beam new_candidate = groups[group_idx].beams[beam_idx]` and update it with new values? 
It will look better than `push_back` and then operate with last vector element

Reply:
Applied